source/Ship.cpp

```diff
@@ -3372,39 +3388,75 @@ bool Ship::CanBeCarried() const
 
 
 
-bool Ship::Carry(const shared_ptr<Ship> &ship)
+bool Ship::Carry(const shared_ptr<Ship> &ship, vector<Visual> *visuals)
 {
 	if(!ship || !ship->CanBeCarried())
 		return false;
 
-	// Check only for the category that we are interested in.
-	const string &category = ship->attributes.Category();
+	Bay *dockBay = ship->reservedBay;
+	if(!dockBay)
+	{
+		// The ship to carry has no bay assigned to it. Find one.
 
-	for(Bay &bay : bays)
-		if((bay.category == category) && !bay.ship)
-		{
-			bay.ship = ship;
-			ship->SetSystem(nullptr);
-			ship->SetPlanet(nullptr);
-			ship->SetTargetSystem(nullptr);
-			ship->SetTargetStellar(nullptr);
-			ship->SetParent(shared_from_this());
-			ship->isThrusting = false;
-			ship->isReversing = false;
-			ship->isSteering = false;
-			ship->commands.Clear();
-			// If this fighter collected anything in space, try to store it
-			// (unless this is a player-owned ship).
-			if(!isYours && cargo.Free() && !ship->Cargo().IsEmpty())
-				ship->Cargo().TransferAll(cargo);
-			// Return unused fuel to the carrier, for any launching fighter that needs it.
-			ship->TransferFuel(ship->fuel, this);
-
-			// Update the cached mass of the mothership.
-			carriedMass += ship->Mass();
-			return true;
+		// Check only for the category that we are interested in.
+		const string &category = ship->attributes.Category();
+		for(Bay &bay : bays)
+			if(bay.category == category && bay.free)
+			{
+				dockBay = &bay;
+				break;
+			}
+
+		if(!dockBay)
+			return false;
+	}
+
+	dockBay->ship = ship;
+	dockBay->free = false;
+	ship->reservedBay = dockBay;
+	ship->SetSystem(nullptr);
+	ship->SetPlanet(nullptr);
+	ship->SetTargetSystem(nullptr);
+	ship->SetTargetStellar(nullptr);
+	ship->SetParent(shared_from_this());
+	ship->isThrusting = false;
+	ship->isReversing = false;
+	ship->isSteering = false;
+	ship->commands.Clear();
+	// If this fighter collected anything in space, try to store it
```

Comment:
Do we need to clear the fire commands here, too?

(I.e. a bug we missed in that other PR)